include/drivers/pinctrl.h

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * Public APIs for pin control drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PINCTRL_H_
+#define ZEPHYR_INCLUDE_DRIVERS_PINCTRL_H_
+
+/**
+ * @brief Pin Controller Interface
+ * @defgroup pinctrl_interface Pin Controller Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <device.h>
+#include <devicetree.h>
+#include <devicetree/pinctrl.h>
+#include <sys/util.h>
+
+#if CONFIG_PINCTRL_NRF
+#include "pinctrl/nrf.h"
+#elif defined(CONFIG_PINCTRL_STM32)
+#include "pinctrl/stm32.h"
+#else
+#error No pinctrl driver is enabled
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Pin controller state. */
+enum pinctrl_state {
+	/** Default. */
+	PINCTRL_STATE_DEFAULT,
+	/** Sleep. */
+	PINCTRL_STATE_SLEEP,
+	/** This and higher values refer to custom private states. */
+	PINCTRL_STATE_PRIV_START,
+};
+
+/** Pin control state configuration. */
+struct pinctrl_state_config {
+	/** State identifier. */
+	enum pinctrl_state id;
+	/** Pin configurations. */
+	const pinctrl_pin_t *pins;
+	/** Number of pin configurations. */
+	size_t pins_cnt;
+};
+
+/** Pin controller configuration for a given device. */
+struct pinctrl_config {
+#ifdef CONFIG_PINCTRL_STORE_REG
+	/** Device address. */
+	uintptr_t reg;
+#endif
+	/** List of state configurations. */
+	const struct pinctrl_state_config **states;
```

Comment:
Is it needed that this field points to an array of pointers to structures with state configurations instead of directly to an array of such structures? Is this only a question of initialization of this array or is there another reason for this?

Reply:
That is indeed a good point. It uses an array of pointers to structures due to the way things are initialized. If we use an array of structures, we'd have something like this (haven't tested):

```c
#define Z_PINCTRL_STATE_INIT(state_idx, node_id)			       \
	COND_CODE_1(Z_PINCTRL_SKIP_STATE(state_idx, node_id), (),	       \
	({								       \
		.id = Z_PINCTRL_STATE_ID(state_idx, node_id),		       \
		.pins = Z_PINCTRL_PINS_NAME(state_idx, node_id),	       \
		.pins_cnt = ARRAY_SIZE(Z_PINCTRL_PINS_NAME(state_idx, node_id))\
	},))

#define Z_PINCTRL_STATES_DEFINE(node_id)				       \
	static const struct pinctrl_state_config			       \
	Z_PINCTRL_STATES_NAME(node_id)[] = {				       \
		UTIL_LISTIFY(DT_NUM_PINCTRL_STATES(node_id),		       \
			     Z_PINCTRL_STATE_INIT, node_id)		       \
	};
```

instead of:

```c
#define Z_PINCTRL_STATE_DEFINE(state_idx, node_id)			       \
	COND_CODE_1(Z_PINCTRL_SKIP_STATE(state_idx, node_id), (),	       \
	(Z_PINCTRL_PINS_DEFINE(state_idx, node_id)			       \
	static const struct pinctrl_state_config			       \
	Z_PINCTRL_STATE_NAME(state_idx, node_id) = {			       \
		.id = Z_PINCTRL_STATE_ID(state_idx, node_id),		       \
		.pins = Z_PINCTRL_PINS_NAME(state_idx, node_id),	       \
		.pins_cnt = ARRAY_SIZE(Z_PINCTRL_PINS_NAME(state_idx, node_id))\
	};))

#define Z_PINCTRL_STATES_DEFINE(node_id)				       \
	UTIL_LISTIFY(DT_NUM_PINCTRL_STATES(node_id),			       \
		     Z_PINCTRL_STATE_DEFINE, node_id)			       \
	static const struct pinctrl_state_config			       \
	*Z_PINCTRL_STATES_NAME(node_id)[] = {				       \
		UTIL_LISTIFY(DT_NUM_PINCTRL_STATES(node_id),		       \
			     Z_PINCTRL_STATES_INIT, node_id)		       \
	};
```

I think the only requirement would be to move `Z_PINCTRL_PINS_DEFINE` out from the state macros, for example:
```c
#define PINCTRL_DT_DEFINE(node_id)					       \
	UTIL_LISTIFY(DT_NUM_PINCTRL_STATES(node_id),			       \
		     Z_PINCTRL_PINS_DEFINE, node_id)			       \
	Z_PINCTRL_STATES_DEFINE(node_id)				       \
	Z_PINCTRL_CONFIG_ATTR struct pinctrl_config			       \
	Z_PINCTRL_CONFIG_NAME(node_id) = Z_PINCTRL_CONFIG_INIT(node_id);
```

Regarding runtime, we now have:

```c
PINCTRL_DT_STATE_DEFINE(DT_PATH(zephyr_user), uart1_alt0_default,
			PINCTRL_STATE_DEFAULT);
#ifdef CONFIG_PM_DEVICE
PINCTRL_DT_STATE_DEFINE(DT_PATH(zephyr_user), uart1_alt0_sleep,
			PINCTRL_STATE_SLEEP);
#endif

const struct pinctrl_state_config *uart1_alt0[] = {
	&uart1_alt0_default,
#ifdef CONFIG_PM_DEVICE
	&uart1_alt0_sleep,
#endif
};
```

With this new approach, we'd need something like:
```c
PINCTRL_DT_PINS_DEFINE(DT_PATH(zephyr_user), uart1_alt1_default);
#ifdef CONFIG_PM_DEVICE
PINCTRL_DT_PINS_DEFINE(DT_PATH(zephyr_user), uart1_alt1_sleep);
#endif

const struct pinctrl_state_config uart1_alt1[] = {
	PINCTRL_DT_STATE_INIT(uart1_alt1_default, PINCTRL_STATE_DEFAULT),
#ifdef CONFIG_PM_DEVICE
	PINCTRL_DT_STATE_INIT(uart1_alt1_sleep, PINCTRL_STATE_SLEEP),
#endif
};
```

Opinions?

Reply:
Yes, good point. The whole structure is gonna be in RAM anyway, so you don't need a double pointer indeed.

Reply:
If we don't want to allow runtime modifications of particular states, only the whole set of them, I think we should get rid of this intermediary array of pointers, as I don't see any value it adds while it costs some flash (and cycles).
Regarding runtime, what about trying to use in the application the same initialization macros as used by drivers? I mean, if we could have a node with some compatible that includes properties from `pinctrl-device.yaml` and define alternative states in that node, then the structures could be initialized the same way as in the driver. I used the following hack just to try if it would be possible:
```
/ {
	uart1_alt: fake_uart {
		compatible = "nordic,nrf-uarte";
		status = "disabled";

		pinctrl-0 = <&uart1_alt1_default>;
		pinctrl-1 = <&uart1_alt1_sleep>;
		pinctrl-names = "default", "sleep";
	};
};
```
an then in the application I used:
```
Z_PINCTRL_STATES_DEFINE(DT_NODELABEL(uart1_alt));
#define UART1_ALT_STATES  Z_PINCTRL_STATES_NAME(DT_NODELABEL(uart1_alt))
```
and:
```
pinctrl_states_configure(uart1_config,
	UART1_ALT_STATES, ARRAY_SIZE(UART1_ALT_STATES));
```
It seems to work. And the `zephyr,user` node is not needed then.

Reply:
> If we don't want to allow runtime modifications of particular states, only the whole set of them, I think we should get rid of this intermediary array of pointers, as I don't see any value it adds while it costs some flash (and cycles).

:+1: 

> Regarding runtime, what about trying to use in the application the same initialization macros as used by drivers? I mean, if we could have a node with some compatible that includes properties from `pinctrl-device.yaml` and define alternative states in that node, then the structures could be initialized the same way as in the driver. I used the following hack just to try if it would be possible:
> 
> ```
> / {
> 	uart1_alt: fake_uart {
> 		compatible = "nordic,nrf-uarte";
> 		status = "disabled";
> 
> 		pinctrl-0 = <&uart1_alt1_default>;
> 		pinctrl-1 = <&uart1_alt1_sleep>;
> 		pinctrl-names = "default", "sleep";
> 	};
> };
> ```
> 

Sounds like a powerful alternative, let's discuss it.

Reply:
@anangl I've changed to an array of structures